pages/06-higher-order-functions.md

```diff
@@ -285,213 +242,180 @@ doing something with those pairs or getting a set of solutions and
-"GAYBALLS"
-~~~~
-
-All of this could also be achived with list comprehensions by the use of
-predicates. There's no set rule for when to use map and filter versus
-using list comprehension, you just have to decide what's more readable
-depending on the code and the context. The filter equivalent of applying
-several predicates in a list comprehension is either filtering something
-several times or joining the predicates with the logical && function.
-
-Remember our quicksort function from the [previous chapter](recursion)?
-We used list comprehensions to filter out the list elements that are
-smaller than (or equal to) and larger than the pivot. We can achieve the
-same functionality in a more readable way by using filter:
-
-~~~~ {.haskell:ghci name="code"}
-quicksort :: (Ord a) => [a] -> [a]
-quicksort [] = []
-quicksort (x:xs) =
-    let smallerSorted = quicksort (filter (<=x) xs)
-        biggerSorted = quicksort (filter (>x) xs)
-    in  smallerSorted ++ [x] ++ biggerSorted
-~~~~
+```elm
+> List.filter ((>) 3) [1,5,3,2,1,6,4,3,2,1]
+[5,6,4] : List number
+> List.filter ((==) 3) [1,2,3,4,5]
+[3] : List number
+> List.filter (let nonEmpty ls = not (List.isEmpty ls) in nonEmpty) \
+|    [[1,2,3],[],[3,4,5],[2,2],[],[],[]]
+[[1,2,3],[3,4,5],[2,2]] : List (List number)
+```
 
 ![map](img/map.png)
 
 Mapping and filtering is the bread and butter of every functional
-programmer's toolbox. Uh. It doesn't matter if you do it with the map
-and filter functions or list comprehensions. Recall how we solved the
-problem of finding right triangles with a certain circumference. With
+programmer's toolbox. Recall how we solved the problem of finding right
+triangles with a certain circumference. With
 imperative programming, we would have solved it by nesting three loops
 and then testing if the current combination satisfies a right triangle
 and if it has the right perimeter. If that's the case, we would have
 printed it out to the screen or something. In functional programming,
 that pattern is achieved with mapping and filtering. You make a function
 that takes a value and produces some result. We map that function over a
 list of values and then we filter the resulting list out for the results
-that satisfy our search. Thanks to Haskell's laziness, even if you map
-something over a list several times and filter it several times, it will
-only pass over the list once.
+that satisfy our search.
 
 Let's *find the largest number under 100,000 that's divisible by 3829*.
 To do that, we'll just filter a set of possibilities in which we know
 the solution lies.
 
-~~~~ {.haskell:hs name="code"}
-largestDivisible :: (Integral a) => a
-largestDivisible = head (filter p [100000,99999..])
-    where p x = x `mod` 3829 == 0
-~~~~
+```elm
+largestDivisible : Maybe Int
+largestDivisible =
+    let
+        p x = x % 3829 == 0
+    in
+    List.head (List.filter p List.reverse (List.range 0 99999))
```

Comment:
No laziness, so some of the examples translated a bit awkwardly (like here, where we had to create a list that was larger than we strictly would have needed.